### PR TITLE
Stories/state public adjustments

### PR DESCRIPTION
**Pull request by eanders**

### CODE REVIEW

File: app/models/grda_warehouse/shape/county.rb [hunk 1]
```diff
@@ -27,10 +28,18 @@ def name
...
 
       def self.zip_codes
+        ZipCode.joins(<<~SQL)
+          JOIN shape_counties ON (
+            LOWER(shape_counties.namelsad) = shape_zip_codes.county_name_lower
+          )
+        SQL
+      end
+
+      def self.spacial_zip_codes
```
Comment (yakloinsteak) [inline app/models/grda_warehouse/shape/county.rb]: probably my original spelling mistake. should be "spatial"

**Review by yakloinsteak — APPROVED** [review]
Looks good to me.

File: app/models/grda_warehouse/shape/county.rb (hunk 1 above)
Comment (eanders) [reply app/models/grda_warehouse/shape/county.rb]: I think that was me, I knew it didn't look right.  Thanks!

File: app/models/grda_warehouse/data_source.rb [hunk 2]
```diff
@@ -122,19 +122,27 @@ class GrdaWarehouse::DataSource < GrdaWarehouseBase
   end
 
   def self.source_data_source_ids
-    GrdaWarehouse::DataSource.source.pluck(:id)
+    Rails.cache.fetch(__method__, expires_in: 1.hours) do
```
Comment (awisema) [inline app/models/grda_warehouse/data_source.rb:126]: Add
```
[:source_data_source_ids, :destination_data_source_ids, :authoritative_data_source_ids, :window_data_source_ids].each { |key| Rails.cache.delete(key) }
```
to an `after_create` hook to force expire the caches?

File: drivers/public_reports/app/models/public_reports/state_level_homelessness.rb [hunk 3]
```diff
@@ -700,11 +705,12 @@ def map_colors
     end
 
     private def homeless_chart_breakdowns(section_title:, charts:, setup:, scope:, date:)
+      iso_date = date.iso8601
       ch_t = GrdaWarehouse::HudChronic.arel_table
-      chronic_date = GrdaWarehouse::HudChronic.where(ch_t[:date].lteq(end_date)).maximum(:date)
+      chronic_date = GrdaWarsehouse::HudChronic.where(ch_t[:date].lteq(end_date)).maximum(:date)
```
Comment (awisema) [inline drivers/public_reports/app/models/public_reports/state_level_homelessness.rb]:
```suggestion
      chronic_date = GrdaWarehouse::HudChronic.where(ch_t[:date].lteq(end_date)).maximum(:date)
```

**Review by awisema — CHANGES_REQUESTED** [review]

File: drivers/public_reports/app/models/public_reports/state_level_homelessness.rb [hunk 4]
```diff
@@ -135,6 +135,122 @@ def household_types
...
+        under_threshold
+      when 'unsheltered_percent'
+        unsheltered_count = data['unsheltered_clients'].to_f || 0.0
+        sheltered_count = data['homeless_clients'] || 0
+        percent = if unsheltered_count.zero? || sheltered_count.zero?
+          0
+        elsif unsheltered_count > 100 && sheltered_count > 100
+          ((unsheltered_count / sheltered_count) * 100).round
+        else
+          ((unsheltered_count / sheltered_count) * 100).round(10)
```
Comment (awisema) [inline drivers/public_reports/app/models/public_reports/state_level_homelessness.rb]:
```suggestion
          ((unsheltered_count / sheltered_count) * 100).round(-1)
```

**Review by awisema — CHANGES_REQUESTED** [review]

File: app/models/grda_warehouse/hud/client.rb [hunk 5]
```diff
@@ -496,6 +496,14 @@ def hmis_source_visible_by?(user)
       where(Arel.sql('1').in(races.map { |race| c_t[race] }))
     end
 
+    # a scope where no race was chosen
+    # potentially this should include RaceNone: [8, 9, 99], but if all others are 0 or 99, then
+    # we really don't know the race
+    scope :with_race_none, -> do
+      not_race = [0, 99]
+      where(AmIndAKNative: not_race, Asian: not_race, BlackAfAmerican: not_race, NativeHIPacific: not_race, White: not_race)
```
Comment (awisema) [inline app/models/grda_warehouse/hud/client.rb:504]: Do we need `nil in here`?
```suggestion
      query = []
      [:AmIndAKNative, :Asian, :BlackAfAmerican, :NativeHIPacific, :White].each do |key|
        part = c_t[key].in([0, 99].or(c_t[key].eq(nil)
        query = query.present? ? query.and(part) : part
      end
      where(query)
```

File: drivers/public_reports/app/views/public_reports/warehouse_reports/state_level_homelessness/_homeless_breakdowns.haml [hunk 6]
```diff
@@ -24,20 +24,22 @@
...
 
 :javascript
   $('.jQuarters').on('change', function(e) {
     var index = $(this).val();
     var selected_date = $(this).data('dates')[index];
     $('.jHomelessBreakdownRow').each(function(e){
-      var chronic_percent = $(this).data('chronics')[selected_date];
+      var title = $(this).find('.jChronicPercent').data('category');
+      var chronic_percent = $(this).data('chronics')[selected_date][title];
+      console.log(title);
```
Comment (awisema) [inline drivers/public_reports/app/views/public_reports/warehouse_reports/state_level_homelessness/_homeless_breakdowns.haml]: Comment this out?

File: drivers/public_reports/app/views/public_reports/warehouse_reports/state_level_homelessness/_map_js.haml [hunk 7]
```diff
@@ -90,18 +90,11 @@
...
-      let division_size = this.max_rate / this.colors.length
-      let division = Math.floor(rate / division_size)
-      if(division == 0) {
-        if(rate > 0) {
-          return this.colors[1]
-        } else {
-          return this.colors[0]
+      for(let color in this.colors) {
+        if(this.colors[color].low <= rate && this.colors[color].high >= rate) {
+          console.log(color, rate, this.colors[color].low, this.colors[color].high)
```
Comment (awisema) [inline drivers/public_reports/app/views/public_reports/warehouse_reports/state_level_homelessness/_map_js.haml]: Comment this out?

File: drivers/public_reports/app/views/public_reports/warehouse_reports/state_level_homelessness/_map_place_js.haml [hunk 8]
```diff
@@ -91,18 +91,11 @@
...
-      let division_size = this.max_rate / this.colors.length
-      let division = Math.floor(rate / division_size)
-      if(division == 0) {
-        if(rate > 0) {
-          return this.colors[1]
-        } else {
-          return this.colors[0]
+      for(let color in this.colors) {
+        if(this.colors[color].low <= rate && this.colors[color].high >= rate) {
+          console.log(color, rate, this.colors[color].low, this.colors[color].high)
```
Comment (awisema) [inline drivers/public_reports/app/views/public_reports/warehouse_reports/state_level_homelessness/_map_place_js.haml]: This too?

File: drivers/public_reports/app/views/public_reports/warehouse_reports/state_level_homelessness/_map_zip_js.haml [hunk 9]
```diff
@@ -93,18 +93,11 @@
...
-      let division_size = this.max_rate / this.colors.length
-      let division = Math.floor(rate / division_size)
-      if(division == 0) {
-        if(rate > 0) {
-          return this.colors[1]
-        } else {
-          return this.colors[0]
+      for(let color in this.colors) {
+        if(this.colors[color].low <= rate && this.colors[color].high >= rate) {
+          console.log(color, rate, this.colors[color].low, this.colors[color].high)
```
Comment (awisema) [inline drivers/public_reports/app/views/public_reports/warehouse_reports/state_level_homelessness/_map_zip_js.haml]: Same

File: drivers/public_reports/app/models/public_reports/state_level_homelessness.rb [hunk 10]
```diff
@@ -220,28 +336,36 @@ def parsed_pre_calculated_data
...
-          end
-          m_colors[color] = { description: description, range: (division_start..division_end) }
-        end
+
+        m_colors[colors[0]] = { description: '0%', range: (0..0), low: 0, high: 0 }
+        m_colors[colors[1]] = { description: 'Any - 10%', range: (0.000001..10.0), low: 0.000001, high: 10.0 }
+        m_colors[colors[2]] = { description: '11% - 15%', range: (10.000001..15.0), low: 10.000001, high: 15.0 }
+        m_colors[colors[3]] = { description: '15% - 20%', range: (15.000001..20.0), low: 15.000001, high: 20.0 }
+        m_colors[colors[4]] = { description: '20% - 25%', range: (20.000001..25.0), low: 20.000001, high: 25.0 }
+        m_colors[colors[5]] = { description: '25%+', range: (25.000001..100.0), low: 25.000001, high: 100.0 }
```
Comment (awisema) [inline drivers/public_reports/app/models/public_reports/state_level_homelessness.rb:354]: Do these match up with the rounding functions?

**Review by awisema — CHANGES_REQUESTED** [review]

File: app/models/grda_warehouse/hud/client.rb (hunk 5 above)
Comment (eanders) [reply app/models/grda_warehouse/hud/client.rb:504]: These aren't null-able per the spec

File: drivers/public_reports/app/models/public_reports/state_level_homelessness.rb (hunk 10 above)
Comment (eanders) [reply drivers/public_reports/app/models/public_reports/state_level_homelessness.rb:354]: They should be granular enough to catch what we need.

**Pull request merged**